File: karaoke/api/routes/queue.py
from karaoke.services.queue_service import QueueService

_queue = QueueService()


async def enqueue(song_id: int):
    return await _queue.enqueue(song_id)


async def list_pending():
    return await _queue.list_pending()


async def list_history(page: int = 1):
    return await _queue.list_history(page)


async def list_usually(page: int = 1):
    return await _queue.list_usually(page)


async def set_top(song_id: int):
    return await _queue.set_top(song_id)


async def remove(song_id: int):
    return await _queue.remove(song_id)

File: crates/api/src/routes/queue.rs
//! 点歌队列路由。对应 Python `karaoke/api/routes/queue.py`。

use crate::response::ApiJson;
use crate::state::AppState;
use axum::extract::{Path, Query, State};
use serde::Deserialize;

pub async fn enqueue(State(state): State<AppState>, Path(song_id): Path<i64>) -> ApiJson {
    state.services.queue.enqueue(song_id).await.into()
}

pub async fn list_pending(State(state): State<AppState>) -> ApiJson {
    state.services.queue.list_pending().await.into()
}

#[derive(Deserialize)]
pub struct PageQuery {
    #[serde(default = "default_page")]
    page: i64,
}

fn default_page() -> i64 {
    1
}

pub async fn list_history(State(state): State<AppState>, Query(q): Query<PageQuery>) -> ApiJson {
    state.services.queue.list_history(q.page).await.into()
}

pub async fn list_usually(State(state): State<AppState>, Query(q): Query<PageQuery>) -> ApiJson {
    state.services.queue.list_usually(q.page).await.into()
}

pub async fn set_top(State(state): State<AppState>, Path(song_id): Path<i64>) -> ApiJson {
    state.services.queue.set_top(song_id).await.into()
}

pub async fn remove(State(state): State<AppState>, Path(song_id): Path<i64>) -> ApiJson {
    state.services.queue.remove(song_id).await.into()
}
